File: java/MarsRovers.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collection;

public class MarsRovers {
	private Plateau plateau;
	Collection<RoverWithInstructions> roversWithInstructions 
		= new ArrayList<RoverWithInstructions>();

	class RoverWithInstructions {
		public RoverWithInstructions(Rover rover, String instructions) {
			this.rover = rover;
			this.instructions = instructions;
		}
		public Rover rover;
		public String instructions;
	}
	
	public MarsRovers(int maxX, int maxY) {
		plateau = new Plateau(maxX, maxY);
	}
	
	public void addRoversData(int x, int y, char heading, String instructions) 
		throws Exception {
		Rover rover = new Rover(x, y, heading, plateau);
		roversWithInstructions.add(new RoverWithInstructions(rover, instructions));
	}
	
	public Collection<Rover> getRovers() {
		Collection<Rover> rovers = new ArrayList<Rover>();
		for ( RoverWithInstructions roverWithInstrictions : roversWithInstructions ) {
			rovers.add(roverWithInstrictions.rover);
		}
		return rovers;
	}
	
	public void run() {
		for ( RoverWithInstructions roverWithInstrictions : roversWithInstructions ) {
			try {
				roverWithInstrictions.rover.execute(roverWithInstrictions.instructions);
			} catch (Exception e) {
				e.printStackTrace();
			}
		}
	}
	
	/**
	 * @param args
	 * @throws Exception 
	 */
	public static void main(String[] args) throws Exception {
		BufferedReader stdin = new BufferedReader(new InputStreamReader(System.in));
		
		// read plateau data and instantiate simulation class
		String plateauXY = stdin.readLine();
		String[] plateauXYArray = plateauXY.split(" ");
		int plateauX = Integer.parseInt(plateauXYArray[0]);
		int plateauY = Integer.parseInt(plateauXYArray[1]);
		MarsRovers marsRovers = new MarsRovers(plateauX, plateauY);

		// read rovers with data
		String line;
		while ((line = stdin.readLine()) != null) {
			String instructions = stdin.readLine();
			
			String[] roverLine = line.split(" ");
			int x = Integer.parseInt(roverLine[0]);
			int y = Integer.parseInt(roverLine[1]);
			char heading = (roverLine[2].toCharArray())[0];
			
			marsRovers.addRoversData(x, y, heading, instructions);
		}
		
		// run the simulation
		marsRovers.run();
		
		for (Rover rover : marsRovers.getRovers()) {
			System.out.println(rover);
		}
	}

}

File: scala/MarsRovers.scala
import scala.io.Source

object MarsRovers {
    def simulate(file: Option[String]) {
        val input = file match {
        	case Some(file) => Source.fromFile(file)
        	case None => Source.fromInputStream(System.in)
        }

        val lines = input.getLines
        val plateauXY = lines.next.trim.split("\\s")
        val plateau = new Plateau(plateauXY(0).toInt, plateauXY(1).toInt)

        val rovers = lines grouped 2 withPartial false map { roverInfo =>
            val (start, instructions) = (roverInfo(0), roverInfo(1))
            val (x, y, h) = parseRover(start)
            (new Rover(x, y, h, plateau), instructions.trim)
        }

        // run the simulation
        val roversAfter = rovers.map { rover: (Rover, String) =>
            rover._1.execute(rover._2)
        }

        roversAfter.foreach(println);
    }

    def parseRover(start: String): (Int, Int, Char) = {
        val roverInfo = start.trim.split("\\s")
        val x = roverInfo(0).toInt
        val y = roverInfo(1).toInt
        val heading = roverInfo(2)(0)
        (x, y, heading)
    }

    def main(args: Array[String]) {
    	val fileArg = if (args.length > 0) Some(args(0)) else None  
        simulate(fileArg)
    }
}


